### ISSUE TITLE
`sam validate` fails when running against a SAM docker image

### ISSUE THREAD

**Issue by xirkus** [aws/aws-sam-cli#2613]
### Description:
The AWS quick start template for a docker image fails `sam validate`.

### Steps to reproduce:
1. Run `sam init`.
2. Select `2 - Image (artifact is an image uploaded to an ECR image repository)` for the package type.
3. Select a base image (I use `5 - amazon/python3.7-base` for my example).
4. Name your application
5. `cd` into your application directory and run `sam build`
6. After the build completes run `sam validate`

### Observed result:

```
% sam validate
Template provided at '/path/to/repo/sam/sam-app/template.yaml' was invalid SAM Template.
Error: [InvalidResourceException('HelloWorldFunction', "'ImageUri' must be set.")] ('HelloWorldFunction', "'ImageUri' must be set.")
```

### Expected result:
`sam validate` should be aware when a Docker image is being used as the deployment package type an validate cleanly. For example, the example project generated will successfully deploy with `sam build`.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: MacOS 11.2.1
2. SAM CLI, version 1.17.0
3. AWS region: us-east-1

**Comment by xirkus** [comment]
NVM. I just realized that the default `template.yaml` does not supply a ECR. Perhaps consider adding this resource:

```
FunctionName:
    Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
    Properties:
      PackageType: Image
      ImageUri: !Ref FunctionECR
      Events:
        HelloWorld:
          Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
          Properties:
            Path: /function
            Method: get
    Metadata:
      DockerTag: python3.7-v1
      DockerContext: ./function
      Dockerfile: Dockerfile
```


It has the advantage of require less setup for starting devs and an ECR that is automatically named.

**Comment by xirkus** [comment]
Hmmm. This seems to be a chicken-egg problem. If you specify an ECR resource in your `template.yaml` it seems that `sam deploy` is unable to recognize it and prompts for `Image Repository for FunctionName:`. I think this fundamentally describes a workflow where an ECR repo is treated like a bucket (which is unconventional as each Docker registry should be associated with a single versioned artifact instead of multiple resources, such as different lambdas, being referenced by tag.

Not sure I agree with this design decision.

**Comment by xirkus** [comment]
This behavior is even stranger given that SAM automatically creates a CLISourceBucket and Lambda policy/roles.

**Comment by sriram-mv** [comment]
@xirkus you need to point sam validate to the built template.yaml, which is present under `.aws-sam/build/template.yaml`.

sam package and sam deploy automatically look for the built template under `.aws-sam/build/template.yaml`. However, `sam validate` does not.

```
sam build

sam validate --template-file .aws-sam/build/template.yaml
.../.aws-sam/build/template.yaml is a valid SAM Template
```

This will showcase that its a valid template.

**Comment by xirkus** [comment]
That's interesting (and seems to work), but the documentation clearly states that it looks in your current directory:

```
  -t, --template-file, --template PATH
                                  AWS SAM template file.  [default:
                                  template.[yaml|yml]]
```

If a non-docker template is used, `sam validate` works without any additional arguments (and presumably looks in the current directory for the `template.yaml`.

Are users expected to know that validation looks in the .aws-sam artifact directory?

**Comment by sriram-mv** [comment]
That is a fair callout, this is due to an implementation detail where for Packagetype Image, there are necessary fields on the SAM template which are not yet present before a build occurs, such as `ImageUri`. once built there is a value for that field. 

However, this can be made transparent to the users, if we automatically look at the .aws-sam directory as well before looking in the current directory.

**Comment by xirkus** [comment]
thanks @sriram-mv!

**Comment by xazhao** [comment]
It should be fixed in our latest release 1.30.0. Closing this issue.
